Add async support

kurtjd opened the pull request:
This PR introduces async support to the driver while still preserving sync functionality via the `maybe-async` crate. The sync version can be built via:

`cargo build --features is_sync`

And the async version is built by default. Tests were also updated to support async via the `tokio` async runtime.

`linux-embedded-hal` does not currently seem to support async so the linux example will panic without the `is_sync` feature enabled.

eldruin reviewed (commented):
Thank you for this work!
Traditionally the sync version is the default and async is often chosen via an `async` feature.
Could you explain the reason to default to async here?
Once this is settled, an entry to the changelog should be added.
The MSRV needs to upgraded to Rust 1.75.0 (readme, CI, changelog) due to async as well.

kurtjd commented:
> Thank you for this work! Traditionally the sync version is the default and async is often chosen via an `async` feature. Could you explain the reason to default to async here? Once this is settled, an entry to the changelog should be added. The MSRV needs to upgraded to Rust 1.75.0 (readme, CI, changelog) due to async as well.

Thanks for taking a look! So, the only real reason for default to async is that it appears the `maybe-async` crate defaults to async, and therefore an `is_sync` feature here maps directly to `is_sync` in `maybe-async`. But you make a good point, so I'll rework it to be sync by default.

kurtjd commented:
I've modified it so the `sync` version is now built by default. The `async` version can be built via:
`cargo build --no-default-features`

Ideally, I would have preferred explicitly enabling an `async` feature, but various experimentations with that proved much less ergonomic than this unfortunately.

Other than async support, I've:
- Changed version from `1.0.1` to `1.1.0` in Cargo.toml and CHANGELOG to reflect addition of new, backwards compatible feature
- Updated MSRV in README, CHANGELOG, and CI
- Added a line to the README showing how to build with async enabled (additional async examples I feel would just add clutter and not much value)
- Added entry to CHANGELOG

Please let me know if you would like any other modifications to the PR. Thanks!

eldruin commented on CHANGELOG.md:
```diff
@@ -8,6 +8,14 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 <!-- next-header -->
 ## [Unreleased] - ReleaseDate
 
+## [1.1.0] - 2025-05-07
+
```
```suggestion
```
Let's do this during the release

eldruin commented on CHANGELOG.md:
```diff
@@ -51,7 +59,8 @@ may be some API changes in the future. All changes will be documented in this
 
 <!-- next-url -->
-[Unreleased]: https://github.com/eldruin/tmp1x2-rs/compare/v1.0.1...HEAD
+[Unreleased]: https://github.com/eldruin/tmp1x2-rs/compare/v1.1.0...HEAD
+[1.1.0]: https://github.com/eldruin/tmp1x2-rs/compare/v1.0.1...v1.1.0
```
```suggestion
[Unreleased]: https://github.com/eldruin/tmp1x2-rs/compare/v1.0.1...HEAD
```

eldruin commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "tmp1x2"
-version = "1.0.1"
+version = "1.1.0"
```
```suggestion
version = "1.0.1"
```

eldruin reviewed (commented):
Thank you so much!
As a user, it sounds counter-intuitive to deactivate features in order to enable async support, though.
Could you explain what the problems are with the `async` feature approach?
[Here is an example](https://github.com/eldruin/tcs3472-rs) of doing it that way. It uses the `maybe-async-cfg` crate, though, which is a redesign of `maybe-async`.
Apologies for not having linked to that earlier.
Please let me know what you think.

kurtjd commented:
> Thank you so much! As a user, it sounds counter-intuitive to deactivate features in order to enable async support, though. Could you explain what the problems are with the `async` feature approach? [Here is an example](https://github.com/eldruin/tcs3472-rs) of doing it that way. It uses the `maybe-async-cfg` crate, though, which is a redesign of `maybe-async`. Apologies for not having linked to that earlier. Please let me know what you think.

The main issue I was running into is that `maybe-async` is async by default, and the `is_sync` feature must be enabled manually. This makes it tricky to have `is_sync` enabled by default in this crate because an `async` feature couldn't then go and disable `is_sync` in `maybe-async` due to the way features work in Rust AFAIK.

I had looked at `maybe-async-cfg` briefly but seems like it could be the better crate for this. I'll work on a redesign using that crate and update this PR shortly. Thanks again for the feedback!

kurtjd committed:
Add async support

kurtjd commented:
@eldruin I've rewritten this to make use of `maybe_async_cfg` mostly following your example. Async support can now be enabled via:

`cargo build --features async`

while sync support is left as default.

eldruin reviewed (approved):
Looks great to me, thank you so much!

eldruin commented:
I have published this in [version 1.1.0](https://github.com/eldruin/tmp1x2-rs/releases/tag/v1.1.0) :tada:
